---
title: "IBD calculations using statgenIBD"
author: "Bart-Jan van Rossum"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: false
    number_sections: true
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{IBD calculations using statgenIBD}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.dim = c(7, 4)
)
library(statgenIBD)
op <- options(width = 90)
```

# The statgenIBD package {.unnumbered}

The `statgenIBD` package is developed as an easy-to-use package for Identity By Descent (IBD) calculations for most common populations used in plant breeding.

This vignette describes how to perform these calculations and how to interpret and use its outputs. This will be done using a number of example data sets, both real life and simulated.

# Theory

The calculations are based on Hidden Markov Models (HMM)...[here we can use or refer to PNAS 2011 paper AMPRIL population. Also refer to Lander and Green?]

# Population types

In the package IBD probabilities can be calculated for many different types of populations. In the following table all supported populations are listed. Note that the value of x in the population types is variable, with its maximum value depicted in the last column.

| **Population type** | **Cross**  | **Description**                                            | **max. x** |
|---------------------|------------|------------------------------------------------------------|------------|
| DH                  | biparental | doubled haploid population                                 |            |
| Fx                  | biparental | Fx population (F1, followed by x-1 generations of selfing) | 8          |
| FxDH                | biparental | Fx, followed by DH generation                              | 8          |
| BCx                 | biparental | backcross, second parent is recurrent parent               | 9          |
| BCxDH               | biparental | BCx, followed by DH generation                             | 9          |
| BC1Sx               | biparental | BC1, followed by x generations of selfing                  | 7          |
| BC1SxDH             | biparental | BC1, followed by x generations of selfing and DH           | 6          |
| C3                  | three-way  | three way cross: (AxB) x C                                 |            |
| C3DH                | three-way  | C3, followed by DH generation                              |            |
| C3Sx                | three-way  | C3, followed by x generations of selfing                   | 7          |
| C3SxDH              | three-way  | C3, followed by x generations of selfing and DH generation | 6          |
| C4                  | four-way   | four-way cross: (AxB) x (CxD)                              |            |
| C4DH                | four-way   | C4, followed by DH generation                              |            |
| C4Sx                | four-way   | C4, followed by x generations of selfing                   | 6          |
| C4SxDH              | four-way   | C4, followed by x generations of selfing and DH generation | 6          |

# Examples

We will demonstrate the basic functionality of the package using some example data sets. The first example will be for a real life data set for a relatively simple doubled haploid population. To highlight some specifics of more complex populations and crosses, we will use simulated data.

## Steptoe Morex

The first example is the Steptoe Morex data, described by [@Hayes1993]. This is a population of 150 barley doubled haploid lines with parentage Steptoe / Morex. The population was developed for the North American Barley Genome Mapping Project by the Oregon State University Barley Breeding Program. For a full description see the [website](https://wheat.pw.usda.gov/ggpages/SxM/) of the program.

The map and genotypic file for this population are included in the package.

Both the map and the genotypic file should be in a **tab**-delimited format, and has the same format as used in Flapjack, see: [website](http://flapjack.hutton.ac.uk/en/latest/projects_&_data_formats.html#data-sets-maps-and-genotypes). The map file should be a file without a header. It should contain three columns, corresponding to marker name, chromosome number and position on the chromosome in centiMorgan.

```{r inspectMap}
## Read the map and display the first tows.
map <- read.table(system.file("extdata/SxM", "SxM_map.txt", package = "statgenIBD"))
head(map)
```

The genotypic file should a file with a header containing the marker names. The first column should contain the genotypes, starting with the parents, in this case Morex and Steptoe. The name of this column is irrelevant and can even be an empty string as in this example.

```{r inspectLoc}
## Read the genotypic file and display the first rows and columns.
geno <- read.table(system.file("extdata/SxM", "SxM_geno.txt", package = "statgenIBD"),
                   header = TRUE)
head(geno[, 1:5])
```

Using the map and genotypic files we can compute the IBD probabilities using `calcIBD`. This function computes, per genotype and marker, the probability that it descended from either parent. If this information is known from the marker, the probabilities will simply be 0 for one of the parents and 1 for the other. When a marker score is missing from the input data, the probabilities are calculated using neighboring markers.

```{r sxmIBD}
## Compute IBD probabilities for Steptoe Morex.
SxMIBD <- calcIBD(poptype = "DH",
                  locfile = system.file("extdata/SxM", "SxM_geno.txt",
                                        package = "statgenIBD"),
                  mapfile = system.file("extdata/SxM", "SxM_map.txt",
                                        package = "statgenIBD"))

## Print summary.
summary(SxMIBD)
```

The output of the `calcIBD` function is an object of class `IBDprob`. This object is a `list` that consists of five elements:

| element    | class        | description                                                                                                                                                                                                                                            |
|------------|--------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| map        | `data.frame` | The map for the population.                                                                                                                                                                                                                            |
| markers    | `array`      | A three dimensional array with the IBD probabilities. The dimensions of this array are \#markers x \#genotypes x \#parents. The array contains per combination of marker and genotypes the probabilities that it descended from either of the parents. |
| poptype    | `character`  | The population type.                                                                                                                                                                                                                                   |
| parents    | `character`  | The parents.                                                                                                                                                                                                                                           |
| multiCross | `logical`    | An indicator showing whether in this `IBDprob` object multiple crosses where combined. This is always `FALSE` for the output of the `calcIBD` function, but will be `TRUE` when multiple crosses are combined as we will see in the examples.          |

The `summary` function can be used to get a short description of the content of a `IBDprob` object.

Looking at genotype **dh001** and marker **plc** we can see from the genotypic file printed above that it has a value of 2, equal to that of Steptoe. If we now look at the output of `calcIBD` it shows a value of 1 for pSteptoe and a value of 0 for pMorex.

```{r check_dh001_plc}
SxMIBD$markers["plc", "dh001", ]
```

For the combination of marker **ABG053** and genotype **dh146** the genotypic file contained a missing value. Looking at the output we can see that the probability that this marker came from Morex is about 0.48 and the probability that it came from Steptoe is about 0.52 [MB: Maybe we should try to find another example, both probabilities equal to 0.5]

```{r check_dh146_plc}
SxMIBD$markers["ABG053", "dh146", ]
```

### Evaluation positions

When calling the `calcIBD` function without extra parameters, the only positions for which the calculations are made are the positions present in the map. There are two ways to change this. A first option is to specify the parameter `evaldist`. Setting this to a value of e.g. 5, assures that extra evaluation positions are added in such a way that the maximum distance between evaluation points will be 5.

The extra evaluation points will be spread evenly. If we look at the map for the Steptoe Morex example, we see that the distance between the first two markers on chromosome 1, **plc** and **glx**, is 18.7. To assure a maximum distance between evaluation points of 5, 3 extra evaluation points will be added. Since they will be spread evenly, the distance between them will be 18.7/3 = 4.68. The extra evaluation points will be named EXT_chr_pos, e.g the extra evaluation point at position 4.68 of chromosome 1, will be named EXT_1\_4.68.

```{r sxmIBD_ext}
## Compute IBD probabilities for Steptoe Morex.
## Add extra evaluation positions so positions are at most 5 apart.
SxMIBD_Ext <- calcIBD(poptype = "DH",
                      locfile = system.file("extdata/SxM", "SxM_geno.txt",
                                            package = "statgenIBD"),
                      mapfile = system.file("extdata/SxM", "SxM_map.txt",
                                            package = "statgenIBD"),
                      evaldist = 5)

## Print summary.
summary(SxMIBD_Ext)
```

As the summary shows, the number of evaluation point has increased to 290, from 116 in the original calculation with only the map positions. Looking at the first rows of the map in the output, we see that indeed 3 extra evaluation points have been added between **plc** and **glx**.

```{r sxmIBD_ext_map}
## Show first rows of map in output.
head(SxMIBD_Ext$map)
```

A second way of changing the position for which the computations are made, is by specifying them in a .txt file. This file should be tab-delimited and can be specified in the `calcIBD` function using the `evalposfile` parameter.

In the package a very simple example of such a file is included for the Steptoe Morex data. It specifies three evaluation positions for each of the chromosomes.

```{r inspectPosFile}
## Read the map and display the first tows.
evalposfile <- read.table(system.file("extdata/SxM", "SxM_eval.txt", package = "statgenIBD"))
head(evalposfile)
```

When using a file with evaluation position, IBD calculations are made only for the positions in the file. The information in the mapfile is still used in those computations, but for the marker position themselves the evaluations are not done. The evaluation points will be named EVAL_chr_pos, e.g the first evaluation point in the file, at position 0 of chromosome 1, will be named EVAL_1\_0.

```{r sxmIBD_posFile}
SxMIBD_posFile <- calcIBD(poptype = "DH",
                          locfile = system.file("extdata/SxM", "SxM_geno.txt",
                                                package = "statgenIBD"),
                          mapfile = system.file("extdata/SxM", "SxM_map.txt",
                                                package = "statgenIBD"),
                          evalposfile = system.file("extdata/SxM", "SxM_eval.txt",
                                                    package = "statgenIBD"))

## Print summary.
summary(SxMIBD_posFile)
```

As the summary shows, the number of evaluation points now decreased to 21, 3 for each of the 7 chromosomes.

In case both `evalposfile` and `evaldist` are specified the former takes prevalence and `evaldist` is ignored.

### Extracting value for markers of interest

Often only the subset of markers is of interest for further analysis, e.g. for QTL Mapping of for fitting a multi-QTL model. To extract the computed probabilities for such a marker subset from a `calcIBD` object we can use the `getQTLProb` function.

```{r SxMProbs_one}
## Extract marker probabilities for markers plc and ABG053.
SxM_probs <- getProbs(SxMIBD, markers = c("plc", "ABG053"))
head(SxM_probs)
```

The output of the `getProbs` function is a `data.frame` with a column `geno` containing the genotype and for both markers the probability that it descended from Morex and that it descended from Steptoe.

## Simulated data

Simulated data for several more complex populations and crosses is included in the package. These are very small and just used for the purpose of showing some of the options and output formats in `statgenIBD`.

### F4

First we have a look at an F4 of a two-way RIL population, i.e. an F1 followed by 3 generations of selfing. This population was constructed from parents GREEN and RED. As we can see from the table in the populations section, when doing computations for an F4 populations, we can use `poptype = "Fx"` where x is replaced by 4.

```{r F4IBD}
## Compute IBD probabilities for simulated F4 population.
F4IBD <- calcIBD(poptype = "F4",
                 locfile = system.file("extdata/popF4", "cross.loc",
                                       package = "statgenIBD"),
                 mapfile = system.file("extdata/popF4", "mapfile.map",
                                       package = "statgenIBD"))

## Print summary.
summary(F4IBD)
```

As the summary shows, the output contains probabilities for two parents, GREEN and RED. The heterozygote, having an allele from both parents is called GREENRED.

### C4S3

Also included is a C4S3 population, a four-way cross followed by 3 generations of selfing. The parents in the four-way cross where YELLOW, BLUE, GREEN and RED, with crossing scheme: (YELLOW x BLUE) x (GREEN x RED).

```{r C3S4DHIBD}
## Compute IBD probabilities for simulated C4S3 population.
C4S3IBD <- calcIBD(poptype = "C4S3",
                   locfile = system.file("extdata/popC4S3", "cross.loc",
                                         package = "statgenIBD"),
                   mapfile = system.file("extdata/popC4S3", "mapfile.map",
                                         package = "statgenIBD"))

## Print summary.
summary(C4S3IBD)
```

As the summary shows, the output contains probabilities for four parents, YELLOW, BLUE, GREEN, RED, plus the four possible heterozygote genotypes: YELLOWGREEN, YELLOWRED, BLUEGREEN, BLUERED.

### Multi-cross

The final example shows how IBD computations for multiple populations can be combined. In the example we use simulated data for two F4DH populations, i.e. F4 population as described above followed by a doubled haploid generation. For the first population the parents where A and B, for the second the parents where A and C.

First we compute the IBD probabilities for each of the populations separately.

```{r multiIBD}
## Compute IBD probabilties for AxB.
AB <- calcIBD(poptype = "F4DH",
              locfile = system.file("extdata/multipop", "AxB.loc",
                                    package = "statgenIBD"),
              mapfile = system.file("extdata/multipop", "mapfile.map",
                                    package = "statgenIBD"))

## Print summary.
summary(AB)

## Compute IBD probabilties for Axc.
AC <- calcIBD(poptype = "F4DH",
              locfile = system.file("extdata/multipop", "AxC.loc",
                                    package = "statgenIBD"),
              mapfile = system.file("extdata/multipop", "mapfile.map",
                                    package = "statgenIBD"))

## Print summary.
summary(AC)
```

Now we want to combine the results from both computations for further analyses. To combine to `IBDprob` objects we can use the `c` function. This is only possible if both objects contain the same type of population. Also the evaluation points for both object have to be identical. In this example we have two F4DH populations and we used the same map file in the computations so combining them is fine.

```{r combineIBD}
ABC <- c(AB, AC)
summary(ABC)
```

Looking at the summary, we can see the our combined population now has 180 individuals, 100 from population AB and 80 from population AC. There are also three parents, A, B and C. For all individuals from population AB the probability for parent C will be 0, as will be the probability for parent B for individuals from population AC.

```{r checkMultiIBD}
## Extract probabilities for markers M1_1 and M3_3.
ABCProbs <- getProbs(ABC, markers = c("M1_1", "M3_3"))

## Print probabilities for genotypes AxB0001 and AxC0001.
ABCProbs[ABCProbs$geno %in% c("AxB0001", "AxC0001"), ]
```

Note that the output now contains an extra column cross, indicating the cross the genotype came from.

```{r winddown, include = FALSE}
options(op)
```

------------------------------------------------------------------------

## References
